src/test/java/com/yelp/nrtsearch/server/luceneserver/search/collectors/additional/NestedCollectorOrderTest.java

```diff
@@ -0,0 +1,353 @@
+import com.yelp.nrtsearch.server.grpc.Query;
+import com.yelp.nrtsearch.server.grpc.RangeQuery;
+import com.yelp.nrtsearch.server.grpc.Script;
+import com.yelp.nrtsearch.server.grpc.SearchRequest;
+import com.yelp.nrtsearch.server.grpc.SearchResponse;
+import com.yelp.nrtsearch.server.grpc.TermsCollector;
+import com.yelp.nrtsearch.server.luceneserver.ServerTestCase;
+import com.yelp.nrtsearch.server.luceneserver.script.js.JsScriptEngine;
+import io.grpc.StatusRuntimeException;
+import io.grpc.testing.GrpcCleanupRule;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import org.apache.lucene.index.IndexWriter;
+import org.apache.lucene.index.NoMergePolicy;
+import org.junit.ClassRule;
+import org.junit.Test;
+
+public class NestedCollectorOrderTest extends ServerTestCase {
+  @ClassRule public static final GrpcCleanupRule grpcCleanup = new GrpcCleanupRule();
+
+  protected List<String> getIndices() {
+    return Collections.singletonList(DEFAULT_TEST_INDEX);
+  }
+
+  protected FieldDefRequest getIndexDef(String name) throws IOException {
+    return getFieldsFromResourceFile("/search/collection/nested_order.json");
+  }
+
+  protected void initIndex(String name) throws Exception {
+    IndexWriter writer = getGlobalState().getIndex(name).getShard(0).writer;
+    // don't want any merges for these tests
+    writer.getConfig().setMergePolicy(NoMergePolicy.INSTANCE);
+    getGlobalState()
+        .getIndexStateManager(DEFAULT_TEST_INDEX)
+        .updateLiveSettings(
+            IndexLiveSettings.newBuilder().setSliceMaxSegments(Int32Value.of(1)).build());
+
+    List<AddDocumentRequest> docs = new ArrayList<>();
+    int doc_id = 0;
+    for (int i = 1; i <= 5; ++i) {
+      for (int j = 1; j <= 20; ++j) {
+        AddDocumentRequest request =
+            AddDocumentRequest.newBuilder()
+                .setIndexName(DEFAULT_TEST_INDEX)
+                .putFields(
+                    "doc_id",
+                    MultiValuedField.newBuilder().addValue(Integer.toString(doc_id)).build())
+                .putFields(
+                    "int_field",
+                    MultiValuedField.newBuilder().addValue(Integer.toString(i)).build())
+                .putFields(
+                    "value_field",
+                    MultiValuedField.newBuilder().addValue(Integer.toString(i * j)).build())
+                .putFields(
+                    "value_field_2",
+                    MultiValuedField.newBuilder().addValue(Integer.toString(-i * j)).build())
+                .build();
+        docs.add(request);
+        doc_id++;
+      }
+    }
+
+    // write docs in random order with 10 per segment
+    Collections.shuffle(docs);
+    for (int i = 0; i < 10; ++i) {
+      List<AddDocumentRequest> chunk = new ArrayList<>();
+      for (int j = 0; j < 10; ++j) {
+        chunk.add(docs.get((10 * i) + j));
+      }
+      addDocuments(chunk.stream());
+      writer.commit();
```

Comment:
Nit: I think it'll be better to call commit via grpc api and minimize direct usage of IndexWriter